substrate-archive/src/wasm_tracing.rs

```diff
@@ -0,0 +1,329 @@
+type BlockHash = Vec<u8>;
+
+/// The Event a tracing subscriber collects before sending data to the TracingActor.
+#[derive(Debug)]
+pub struct EventMessage {
+	pub block_num: Option<u32>,
+	pub hash: Option<Vec<u8>>,
+	pub name: String,
+	pub target: String,
+	pub level: Level,
+	pub values: TraceData,
+	pub parent_id: Id,
+	pub time: DateTime<Utc>,
+	pub file: Option<String>,
+	pub line: Option<u32>,
+}
+
+/// The message a tracing subscriber collects before sending data to the TracingActor.
+#[derive(Debug, Clone)]
+pub struct SpanMessage {
+	pub id: Id,
+	pub parent_id: Option<Id>,
+	pub name: String,
+	pub target: String,
+	pub level: Level,
+	pub values: TraceData,
+	pub start_time: DateTime<Utc>,
+	pub overall_time: chrono::Duration,
+	pub file: Option<String>,
+	pub line: Option<u32>,
+}
+
+/// Finished Trace Data Format. Ready for insertion into a relational database.
+#[derive(Debug)]
+pub struct Traces {
+	block_num: u32,
+	hash: Vec<u8>,
+	pub spans: Vec<SpanMessage>,
+	pub events: Vec<EventMessage>,
+}
+
+impl Traces {
+	pub fn new(block_num: u32, hash: Vec<u8>, events: Vec<EventMessage>, spans: Vec<SpanMessage>) -> Self {
+		Traces { block_num, hash, spans, events }
+	}
+
+	/// Get the hash these spans come from.
+	pub fn hash(&self) -> Vec<u8> {
+		self.hash.clone()
+	}
+
+	/// Get the block number of the block these spans come from.
+	pub fn block_num(&self) -> u32 {
+		self.block_num
+	}
+}
+
+#[derive(Debug)]
+pub struct SpanEvents {
+	pub spans: Vec<SpanMessage>,
+	pub events: Vec<EventMessage>,
+}
+
+/// Collects traces and filters based on target.
+/// The Subscriber implementation is blocking. It uses Mutex primitives to coalesce traces before
+/// sending them to the appropriate actor.
+/// Therefore, one must be careful not to block the async executor when adding tracing spans
+/// using this subscriber implementation anywhere inside an async context in substrate-archive.
+pub struct TraceHandler {
+	/// the Block Number this trace set is from
+	block_num: u32,
+	/// Hash of the block for this trace set
+	hash: BlockHash,
+	span_events: Arc<Mutex<SpanEvents>>,
+	targets: Vec<(String, Level)>,
+	counter: AtomicU64,
+	current_span: CurrentSpan,
+}
+
+impl TraceHandler {
+	pub fn new(targets: &str, block_num: u32, hash: BlockHash, span_events: Arc<Mutex<SpanEvents>>) -> Self {
+		let targets = targets.split(',').map(|s| parse_target(s)).collect();
+		// must start indexing from 1 otherwise `tracing` panics
+		let counter = AtomicU64::new(1);
+		Self { targets, counter, span_events, block_num, hash, current_span: Default::default() }
+	}
+
+	/// Formats an event as an [`EventMessage`] and stores it in the [`SpanEvents`] (which is sent to the [`TracingActor`] after the block is executed).
+	fn gather_event(&self, event: &Event<'_>, time: DateTime<Utc>) -> Result<()> {
+		let meta = event.metadata();
+		let mut values = TraceData::default();
+		event.record(&mut values);
+		let parent_id =
```

Comment:
Is there any reason to discard events without a parent?

Reply:
Good catch! This was left over from when I wrapped the entire block execution in a span, so anything outside of that span I considered irrelevant. Now that the subscriber is thread-local that's no longer needed